examples-gallery/plot_Mississippi_steamboat.py

```diff
@@ -0,0 +1,445 @@
+import matplotlib.pyplot as plt
+from matplotlib.animation import FuncAnimation
+from matplotlib.patches import Rectangle
+
+
+# %%
+# Set up the Equations of Motion.
+# -------------------------------
+#
+# Set up the geometry of the system.
+#
+# - :math:`N`: inertial frame of reference
+# - :math:`O`: origin of the inertial frame of reference
+# - :math:`A_S`: body fixed frame of the steamboat
+# - :math:`A_{LW}`: body fixed frame of the left wheel
+# - :math:`A_{RW}`: body fixed frame of the right wheel
+# - :math:`Dmc_S`: center of the steamboat
+# - :math:`Dmc_{LW}`: center of the left wheel
+# - :math:`Dmc_{RW}`: center of the right wheel
+# - :math:`FP_{LW}`: point where the force due to rotation of the left
+#   wheel is applied
+# - :math:`FP_{RW}`: point where the force due to rotation of the right
+#   wheel is applied
+
+N, AS, ALW, ARW = sm.symbols('N, AS, ALW, ARW', cls = me.ReferenceFrame)
+DmcS, DmcLW, DmcRW = sm.symbols('DmcS, DmcLW, DmcRW', cls = me.Point)
+O, FPLW, FPRW = sm.symbols('O, FPLW, FPRW', cls = me.Point)
+
+q, x, y, qLW, qRW = me.dynamicsymbols('q, x, y, qLW, qRW')
+u, ux, uy, uLW, uRW = me.dynamicsymbols('u, ux, uy, uLW, uRW')
+tLW, tRW = me.dynamicsymbols('tLW, tRW')
+
+mS, mW, rW, aS, bS, cS, cW = sm.symbols('mS, mW, rW, aS, bS, cS, cW', real=True)
+
+t = me.dynamicsymbols._t
+O.set_vel(N, 0)
+
+AS.orient_axis(N, q, N.z)
+AS.set_ang_vel(N, u*N.z)
+ALW.orient_axis(AS, qLW, AS.x)
+ALW.set_ang_vel(AS, uLW*AS.x)
+ARW.orient_axis(AS, qRW, AS.x)
+ARW.set_ang_vel(AS, uRW*AS.x)
+
+DmcS.set_pos(O, x*N.x + y*N.y)
+DmcS.set_vel(N, ux*N.x + uy*N.y)
+DmcLW.set_pos(DmcS, -1.1*bS*AS.x)
+DmcLW.v2pt_theory(DmcS, N, AS)
+DmcRW.set_pos(DmcS, 1.1*bS*AS.x)
+DmcRW.v2pt_theory(DmcS, N, AS)
+
+FPLW.set_pos(DmcLW, -rW*N.z)
+FPLW.set_vel(N, DmcLW.vel(N) + uLW*AS.x.cross(-rW*N.z))
+FPRW.set_pos(DmcRW, -rW*N.z)
+FPRW.set_vel(N, DmcRW.vel(N) + uRW*AS.x.cross(-rW*N.z))
+
+
+# %%
+# Set up the drag forces acting on the boat.
+#
+# The drag force acting on a body moving in a fluid is given by
+# :math:`F_D = -c \cdot A \cdot | \bar v|^2 \cdot \dfrac{\bar v}{| \bar v|}`,
+# where :math:`c` is the drag coefficient, :math:`\bar v` is the velocity
+# vector of the body and :math:`A` is the cross section area of the body facing
+# the flow. This may be found here:
+#
+# https://courses.lumenlearning.com/suny-physics/chapter/5-2-drag-forces/
+#
+# In order to avoid numerical issues with .normalize(), sm.sqrt(..)
+# I will use the following:
+#
+# ( :math:`\circ` denotes the inner product)
+#
+# :math:`F_{D_x} = -c \cdot A \cdot (A.x \circ \bar v)^2 \cdot \operatorname{sgn}(A.x \circ \bar v) \cdot A.x`
+# :math:`F_{D_y} = -c \cdot A \cdot (A.y \circ \bar v)^2 \cdot \operatorname{sgn}(A.y \circ \bar v) \cdot A.y`
+#
+# As an (infinitely often) differentiable approximation of the sign function,
+# I will use the fairly standard approximation:
+#
+# :math:`\operatorname{sgn}(x) \approx \tanh( \alpha \cdot x )` with :math:`\alpha \gg 1`
```

Comment:
If you specify all the variables to have correct assumptions (for example `real=True` or `nonegative=True`, then you can use `Abs()` and it will differentiate once and print with the C printers. I don't think you have to use this tangent function.

Reply:
> If you specify all the variables to have correct assumptions (for example `real=True` or `nonegative=True`, then you can use `Abs()` and it will differentiate once and print with the C printers. I don't think you have to use this tangent function.

Mathematically abs(x-a) is not differentiable at x=a, or so I understand it. How does sympy go around this?
(I had a problem with opty and abs before, but I do not think, I declared the variables to be real)

Reply:
```python
In [1]: import sympy as sm

In [2]: t = sm.symbols('t', real=True, nonnegative=True)

In [3]: v = sm.Function('v', real=True)(t)

In [4]: f = sm.Abs(v)

In [5]: f
Out[5]: Abs(v(t))

In [6]: f.diff(t)
Out[6]: sign(v(t))*Derivative(v(t), t)

In [7]: f.diff(v)
Out[7]: sign(v(t))*Derivative(v(t), v(t))
```

Reply:
Thanks!
This means, sympy defines $\dfrac{d}{dt} abs(x(t)) = 0$ at x(t) = 0, it seems.
Would opty not 'prefer' functions which have continuous derivatives, or does this not affect convergence?

Reply:
Probably just leave it as the tanh function.

>  I recall you told me it is better to use t = me.dynamicsymbols._t

That is required if you use the default symbol for t in sympy.physics.mechanics. But you can set t to any variable you want.